Add debug loop execution ordering output

GiudGiud opened the pull request:
- [x] add tests
- [x] make sure FEProblem always knows its execute_on (currently doesnt seem like it)
- [x] make sure we only print the objects that are executed on each execute_on and not all the objects

Features we could consider adding:
- [x] restrict execution output per thread (missing the parameter that's it)
- [x] restrict execution output per block or boundary id

EDIT: per-thread is likely not super helpful for people's dev workflows at this point.
We already have the logic at this point though, just need to add a parameter. 

Currently this looks like this
```
[DBG] Beginning Elemental loop to compute residual on LINEAR
[DBG] Execution order on each element:
[DBG] - kernels on element quadrature points
[DBG] Ordering of kernels on block 0:
[DBG] diff convection diff_v 
```

List of issues where people have been adding this to debug things (found one, feel free to add others)
https://github.com/idaholab/moose/issues/14680

GiudGiud committed:
Add helper routine for output of warehouses
refs #21971

GiudGiud committed:
Add console to Problem to be able to output from loops in console
refs #21971

GiudGiud committed:
Add verbose output to loops
refs #21971

GiudGiud committed:
Add Debug parameter for verbose output of execution order
refs #21971

GiudGiud committed:
Add more loop execution output
Add print of current execute_on flag
Add [DBG] prefix to loop execution output
refs #21971

GiudGiud committed:
Move execution output from Problem to FEProblemBase as Problem isnt accessed directly currently
refs #21971

GiudGiud committed:
Switch to enums for loop execution printing
refs #21971

GiudGiud committed:
Use by-block printing
refs #21971

GiudGiud committed:
Add test for outputting execution order to console
refs #21971

GiudGiud committed:
Add more list of objects executed, clean up a log
refs #21971

GiudGiud committed:
Add a general Debug system documentation and feature new execution ordering printing feature
refs #21971

GiudGiud committed:
Add execution ordering output to FV flux kernels
refs #21971

GiudGiud committed:
Cleaner logs for damping, adaptivity, and avoid printing a bad message for residual, jacobian/both at once
refs #21971

GiudGiud committed:
Add tests for execution logs
refs #21971

GiudGiud committed:
Keep LINEAR exec flag for residual computation
refs #21971

GiudGiud committed:
Only print block ordering once per block
Only print general ordering once per simulation
refs #21971

GiudGiud committed:
Set execute on flag when computing a single user object in problem.
Should the problem always do that?
refs #21971

GiudGiud committed:
Rework FV ordering printing:
- add fviks/fvbcs in a boundary print
- add general log of order on each face
refs #21971

GiudGiud committed:
Add UO execution printing testing on TRANSFER
refs #21971

GiudGiud committed:
Add finite volume execution printing testing
refs #21971

GiudGiud committed:
Print execution ordering only once per block-pair / boundary for finite volume objects
refs #21971

GiudGiud committed:
Rename set of blocks for which execution has been printed
Move code adding to said inside print to keep execution loop lean
- print is no longer a const routine... could use mutable?
refs #21971

GiudGiud committed:
Protect access to active indicator block objects
fixup printing of FV
refs #21971

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1158013/) on 797d49f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9faad45639efa6b11ea6e6ac170d1f39f4d9891e`

GiudGiud committed:
Add routine to print a vector of objects being executed

GiudGiud committed:
Generalize the printing of UO ordering to all threaded element loop : WIP

GiudGiud committed:
Improve formatting in all execution prints:
- add line breaks
- add prefix on all lines

Done in three different ways:
- queries of formatted strings from the warehouse
- printExecutionOrdering for TheadedElementLoop
- ConsoleUtils::formatString on an accumulated string

All 3 rely on same formatting code thankfully

GiudGiud committed:
Enable suffix without the column in formatted prints
Add more execution schedule context
Fix issue that "no UO on this subdomain" was printed several times
Fix regex issues with coloring

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 06b0b74d48cb9ab8152c11db348caa530810a7c9
+Subproject commit d34af5ea41a54755f6cd5ef60150693f582b7f72
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

GiudGiud committed:
Use more compact code for finite volume printing of execution

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1415054/) on e3bc12b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2cd77deee2262685a0c09588f69db50656044f48`

GiudGiud committed:
APA formatting of tests specs

GiudGiud committed:
Only print execution order for thread 0
- would be really easy to select a thread for which to print execution, but with the total lack of use
of threading in the user base, it would also help really little

GiudGiud committed:
Better formatting, less line breaks in the middle of logs, less extra lines

GiudGiud committed:
Fix boundary restricted auxvariable execution output
Fix some typos in test harness
Fix test that tests ordering on something is not consistently ordered: FV BCs

GiudGiud commented:
This should roughly pass.

I d be happy to get feedback on lowering the footprint of this on the loops code. Maybe add some stringify definitions?
I m not sure what to do without making crazy templates that won't really help with code clarity.

GiudGiud commented:
@loganharbour @lindsayad @roystgnr need a volunteer for a coarse first pass

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1419519/) on 1c4f3f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ba109d0eb9dde4fa0965ce7253c956b13bf00549`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1419880/) on 1c7ea0e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21586/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1419915/) on 1c7ea0e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8be85e</th>
      <th colspan=4>#21586 1c7ea0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.88%</td>
      <td>84.84%</td>
      <td>-0.05%</td>
      <td>77.37%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84239</td>
      <td>84612</td>
      <td>+373</td>
      <td>424</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15001</td>
      <td>15121</td>
      <td>+120</td>
      <td>124</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21586/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21586/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21586/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 77.37% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
I think I can get a tad more coverage with simple tests. I ll do That concurrently to addressing reviews

GiudGiud committed:
Fix links in new Debug page
refs #21971

lindsayad commented:
I think I should be able to start reviewing sometime today

lindsayad commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,87 @@
+load/build the mesh, create the variables, kernels, boundary conditions, output objects etc.
+The ordering of this process may be shown using the [!param](/Debug/show_actions) parameter.
+
+Most `Actions` are executed during initialization, but this option is also able to report on the execution of
+`Actions` during the solve.
```
We have actions executing during the `solve`?!

lindsayad commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,87 @@
+### Solve and execution ordering
 
-!! Describe and include an example of how to use the Debug system.
+Nearly every solve in MOOSE consists in a succession of operations on every nodes, quadrature points,
```
```suggestion
Nearly every solve in MOOSE consists of a succession of operations on nodes, quadrature points,
```

lindsayad commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,87 @@
+
+The MOOSE `Debug` system offers the [!param](/Debug/show_execution_order) parameter to output the
+order of execution of each of these objects in those loops. This order may depend on local block/boundary
+restrictions, local or global dependency resolutions.
```
```suggestion
restrictions, and local or global dependency resolutions.
```

lindsayad commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,87 @@
+
+This parameter is most helpful to understand if `AuxKernels`, `UserObjects` and other systems which can
+interact in arbitrarily complex ways on a group of variables are indeed executed in the order desired
+by the modeler. If problematic, object execution may be reordered using various boolean parameters, `execute_on` flags and other manual dependency declarations.
```
```suggestion
by the modeler. If problematic, object execution may be reordered using various boolean parameters, `execute_on` flags, and other manual dependency declarations.
```

lindsayad commented on framework/include/loops/ComputeElemAuxVarsThread.h:
```diff
@@ -39,6 +39,12 @@ class ComputeElemAuxVarsThread : public ThreadedElementLoop<ConstElemRange>
+  // Print list of object types executed and in which order
+  void printGeneralExecutionInformation() const override;
+
+  // Print list of specific objects executed and in which order
```
```suggestion
  /// Print list of specific objects executed and in which order
```

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -112,6 +112,22 @@ class ThreadedFaceLoop
+  virtual void printGeneralExecutionInformation() const {};
+
+  /// Print ordering of objects executed on each block
+  virtual void printBlockExecutionInformation(){};
+
+  /// Print ordering of objects exected on each boundary
+  virtual void printBoundaryExecutionInformation(const BoundaryID /* bnd_id */){};
```
```suggestion
  virtual void printBlockExecutionInformation(){}

  /// Print ordering of objects exected on each boundary
  virtual void printBoundaryExecutionInformation(const BoundaryID /* bnd_id */){}
```

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+void
+ComputeFVFluxThread<RangeType, AttributeTagType>::printBlockExecutionInformation()
+{
+  if (_fe_problem.shouldPrintExecution(_tid) && _fv_flux_kernels.size())
```
Could do early return instead and avoid nesting

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+
+  if (bcs.size())
+  {
+    auto console = _fe_problem.console();
```
```suggestion
    auto & console = _fe_problem.console();
```

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+    auto console = _fe_problem.console();
+    console << "[DBG] FVBCs on boundary " << bnd_id << " between subdomain " << _subdomain
+            << " and neighbor " << _neighbor_subdomain << std::endl;
+    std::string fv_bcs = ConsoleUtils::mooseObjectVectorToString(bcs);
```
```suggestion
    const std::string fv_bcs = ConsoleUtils::mooseObjectVectorToString(bcs);
```

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+  }
+  if (iks.size())
+  {
+    auto console = _fe_problem.console();
```
```suggestion
    auto & console = _fe_problem.console();
```

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+    auto console = _fe_problem.console();
+    console << "[DBG] FVIKs on boundary " << bnd_id << " between subdomain " << _subdomain
+            << " and neighbor " << _neighbor_subdomain << std::endl;
+    std::string fv_iks = ConsoleUtils::mooseObjectVectorToString(iks);
```
```suggestion
    const std::string fv_iks = ConsoleUtils::mooseObjectVectorToString(iks);
```

lindsayad commented on framework/include/loops/ComputeIndicatorThread.h:
```diff
@@ -46,6 +46,12 @@ class ComputeIndicatorThread : public ThreadedElementLoop<ConstElemRange>
+  /// Print general information about the loop when it executes
+  void printGeneralExecutionInformation() const override;
+
+  // Print list of specific objects executed and in which order
```
```suggestion
  /// Print list of specific objects executed and in which order
```

lindsayad commented on framework/include/loops/ComputeJacobianThread.h:
```diff
@@ -77,4 +81,10 @@ class ComputeJacobianThread : public ThreadedElementLoop<ConstElemRange>
   virtual void computeInternalFaceJacobian(const Elem * neighbor);
   virtual void computeInternalInterFaceJacobian(BoundaryID bnd_id);
+
+  // Print list of objects executed and in which order
```
```suggestion
  /// Print list of objects executed and in which order
```

lindsayad commented on framework/include/loops/ComputeResidualAndJacobianThread.h:
```diff
@@ -43,6 +43,12 @@ class ComputeResidualAndJacobianThread : public NonlinearThread
+  /// Print information about the loop, mostly order of execution of objects
+  void printGeneralExecutionInformation() const override;
+
+  // Print list of specific objects executed and in which order
```
```suggestion
  /// Print list of specific objects executed and in which order
```

lindsayad commented on framework/include/loops/ComputeResidualThread.h:
```diff
@@ -33,6 +33,12 @@ class ComputeResidualThread : public NonlinearThread
+  /// Print information about the loop, mostly order of execution of objects
+  void printGeneralExecutionInformation() const override;
+
+  // Print list of specific objects executed and in which order
```
```suggestion
  /// Print list of specific objects executed and in which order
```

lindsayad commented on framework/include/loops/ThreadedElementLoop.h:
```diff
@@ -55,6 +55,22 @@ class ThreadedElementLoop : public ThreadedElementLoopBase<RangeType>
+   * Routine to output the ordering of objects within a vector. These objects must implement the
+   * name() routine, and it must return a string or compatible type.
+   *
+   * @tparam RangeType the type of the element-based loop
```
```suggestion
```
That template parameter is for the class and so should be in the class's doxygen

lindsayad commented on framework/include/loops/ThreadedElementLoop.h:
```diff
@@ -55,6 +55,22 @@ class ThreadedElementLoop : public ThreadedElementLoopBase<RangeType>
+   * objects
+   */
+  template <typename T>
+  void printExecutionOrdering(const std::vector<T *> objs,
```
```suggestion
  void printExecutionOrdering(const std::vector<T *> & objs,
```

lindsayad commented on framework/include/loops/ThreadedElementLoop.h:
```diff
@@ -116,3 +132,28 @@ ThreadedElementLoop<RangeType>::neighborSubdomainChanged()
+template <typename RangeType>
+template <typename T>
+void
+ThreadedElementLoop<RangeType>::printExecutionOrdering(const std::vector<T *> objs,
```
```suggestion
ThreadedElementLoop<RangeType>::printExecutionOrdering(const std::vector<T *> & objs,
```

lindsayad commented on framework/include/loops/ThreadedElementLoop.h:
```diff
@@ -116,3 +132,28 @@ ThreadedElementLoop<RangeType>::neighborSubdomainChanged()
+                                                       const bool print_header,
+                                                       const std::string & line_prefix) const
+{
+  if (objs.size())
+  {
```
```suggestion
  if (!objs.size())
    return;
 
```

lindsayad commented on framework/include/loops/ThreadedElementLoop.h:
```diff
@@ -116,3 +132,28 @@ ThreadedElementLoop<RangeType>::neighborSubdomainChanged()
+    std::vector<MooseObject *> moose_objs;
+    for (auto obj_ptr : objs)
+      moose_objs.push_back(dynamic_cast<MooseObject *>(obj_ptr));
+    auto names = ConsoleUtils::mooseObjectVectorToString(moose_objs);
```
```suggestion
    const auto names = ConsoleUtils::mooseObjectVectorToString(moose_objs);
```

lindsayad commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -166,6 +166,18 @@ class ThreadedElementLoopBase
+  virtual void printGeneralExecutionInformation() const {};
+
+  /// Print information about the particular ordering of objects on each block
+  virtual void printBlockExecutionInformation(){};
```
```suggestion
  virtual void printBlockExecutionInformation(){}
```

lindsayad commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -166,6 +166,18 @@ class ThreadedElementLoopBase
+  std::set<SubdomainID> _blocks_exec_printed;
+
+  /// Resets the set of blocks visited
+  virtual void clearBlocksVisited() { _blocks_exec_printed.clear(); }
```
Does this need to be virtual?

GiudGiud replied:
```suggestion
  void clearBlocksVisited() { _blocks_exec_printed.clear(); }
```

lindsayad commented on framework/include/loops/ThreadedNodeLoop.h:
```diff
@@ -71,6 +71,9 @@ class ThreadedNodeLoop
   THREAD_ID _tid;
+
+  /// Print information about the loop, mostly order of execution of objects
+  virtual void printGeneralExecutionInformation() const {};
```
```suggestion
  virtual void printGeneralExecutionInformation() const {}
```

lindsayad commented on framework/include/outputs/ConsoleUtils.h:
```diff
@@ -90,4 +91,19 @@ std::string outputLegacyInformation(MooseApp & app);
+ * @param objs the vector with all the MooseObjects
+ * @param sep a separator in between each object's name
+ */
+std::string mooseObjectVectorToString(const std::vector<MooseObject *> objs,
```
```suggestion
std::string mooseObjectVectorToString(const std::vector<MooseObject *> & objs,
```

lindsayad commented on framework/include/problems/Problem.h:
```diff
@@ -40,7 +40,12 @@ class Problem : public MooseObject, public PerfGraphInterface
-  virtual bool isSolveTerminationRequested() { return _termination_requested; };
+  virtual bool isSolveTerminationRequested() const { return _termination_requested; };
+
+  /**
+   * Return console handle
+   */
+  const ConsoleStream & console() const { return _console; }
```
```suggestion
  /**
   * Return console handle
   */
  const ConsoleStream & console() const { return _console; }
  
  /**
   * Return console handle
   */
  ConsoleStream & console() { return _console; }  
```

lindsayad commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -743,6 +749,17 @@ MooseObjectWarehouseBase<T>::subdomainsCovered(std::set<SubdomainID> & subdomain
 
+template <typename T>
+std::string
+MooseObjectWarehouseBase<T>::activeObjectsToFormattedString(THREAD_ID tid /*=0*/,
```
```suggestion
MooseObjectWarehouseBase<T>::activeObjectsToFormattedString(const THREAD_ID tid /*=0*/,
```

lindsayad commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -743,6 +749,17 @@ MooseObjectWarehouseBase<T>::subdomainsCovered(std::set<SubdomainID> & subdomain
+template <typename T>
+std::string
+MooseObjectWarehouseBase<T>::activeObjectsToFormattedString(THREAD_ID tid /*=0*/,
+                                                            const std::string & prefix) const
```
```suggestion
                                                            const std::string & prefix /*="[DBG]"*/) const
```

lindsayad commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -743,6 +749,17 @@ MooseObjectWarehouseBase<T>::subdomainsCovered(std::set<SubdomainID> & subdomain
+{
+  std::string output;
+  for (const auto & object : _active_objects[tid])
+    output += object->name() + " ";
```
Should use `join` here I think in order to avoid a final trailing space (unless that's what you want)

GiudGiud replied:
done

lindsayad reviewed (changes requested):
Through the headers

GiudGiud commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,87 @@
+This parameter is most helpful to understand if `AuxKernels`, `UserObjects` and other systems which can
+interact in arbitrarily complex ways on a group of variables are indeed executed in the order desired
+by the modeler. If problematic, object execution may be reordered using various boolean parameters, `execute_on` flags and other manual dependency declarations.
+
```
```suggestion
For example for UserObjects, the `execution_order_group` parameter lets the modeler select the ordering of executions of user objects.

```

GiudGiud commented on framework/include/loops/ComputeElemAuxVarsThread.h:
```diff
@@ -39,6 +39,12 @@ class ComputeElemAuxVarsThread : public ThreadedElementLoop<ConstElemRange>
+  void printGeneralExecutionInformation() const override;
+
+  // Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
nope, because the list of blocks in which a print has happened is updated.

I could also make that list mutable if that s preferred

GiudGiud replied:
I ll just do that

GiudGiud committed:
Apply suggestions from code review

- more const
- more references
- better explanations
- less extra semi-columns
refs #21971
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented on framework/include/loops/ComputeElemAuxBcsThread.h:
```diff
@@ -44,4 +44,11 @@ class ComputeElemAuxBcsThread
+
+  /// Print list of specific objects executed and in which order
+  void printBoundaryExecutionInformation(unsigned int boundary_id,
+                                         std::vector<std::shared_ptr<AuxKernelType>> kernels) const;
```
```suggestion
                                         const std::vector<std::shared_ptr<AuxKernelType>> & kernels) const;
```

GiudGiud commented on framework/include/loops/ComputeElemAuxVarsThread.h:
```diff
@@ -39,6 +39,12 @@ class ComputeElemAuxVarsThread : public ThreadedElementLoop<ConstElemRange>
+  void printGeneralExecutionInformation() const override;
+
+  /// Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -112,6 +112,22 @@ class ThreadedFaceLoop
+  virtual void printGeneralExecutionInformation() const {}
+
+  /// Print ordering of objects executed on each block
+  virtual void printBlockExecutionInformation(){}
```
```suggestion
  virtual void printBlockExecutionInformation() const {}
```

GiudGiud commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -130,6 +146,12 @@ class ThreadedFaceLoop
   SubdomainID _old_neighbor_subdomain;
 
+  /// Set to keep track of blocks for which we have printed the execution pattern
+  std::set<std::pair<const SubdomainID, const SubdomainID>> _blocks_exec_printed;
```
```suggestion
  mutable std::set<std::pair<const SubdomainID, const SubdomainID>> _blocks_exec_printed;
```

GiudGiud commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -112,6 +112,22 @@ class ThreadedFaceLoop
+  virtual void printBlockExecutionInformation(){}
+
+  /// Print ordering of objects exected on each boundary
+  virtual void printBoundaryExecutionInformation(const BoundaryID /* bnd_id */){}
```
```suggestion
  virtual void printBoundaryExecutionInformation(const BoundaryID /* bnd_id */) const {}
```

GiudGiud commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -972,3 +1019,89 @@ ComputeFVFluxRJThread<RangeType>::postFace(const FaceInfo & /*fi*/)
+
+template <typename RangeType, typename AttributeTagType>
+void
+ComputeFVFluxThread<RangeType, AttributeTagType>::printBlockExecutionInformation()
```
```suggestion
ComputeFVFluxThread<RangeType, AttributeTagType>::printBlockExecutionInformation() const
```

GiudGiud commented on framework/include/loops/ComputeIndicatorThread.h:
```diff
@@ -46,6 +46,12 @@ class ComputeIndicatorThread : public ThreadedElementLoop<ConstElemRange>
+  void printGeneralExecutionInformation() const override;
+
+  /// Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ComputeJacobianThread.h:
```diff
@@ -77,4 +81,10 @@ class ComputeJacobianThread : public ThreadedElementLoop<ConstElemRange>
+  void printGeneralExecutionInformation() const override;
+
+  /// Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ComputeMarkerThread.h:
```diff
@@ -38,6 +38,12 @@ class ComputeMarkerThread : public ThreadedElementLoop<ConstElemRange>
+  void printGeneralExecutionInformation() const override;
+
+  /// Print information about ordering of objects on each block
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ComputeResidualAndJacobianThread.h:
```diff
@@ -43,6 +43,12 @@ class ComputeResidualAndJacobianThread : public NonlinearThread
+  void printGeneralExecutionInformation() const override;
+
+  /// Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ComputeResidualThread.h:
```diff
@@ -33,6 +33,12 @@ class ComputeResidualThread : public NonlinearThread
+  void printGeneralExecutionInformation() const override;
+
+  /// Print list of specific objects executed and in which order
+  void printBlockExecutionInformation() override;
```
```suggestion
  void printBlockExecutionInformation() const override;
```

GiudGiud commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -166,6 +166,18 @@ class ThreadedElementLoopBase
+  virtual void printGeneralExecutionInformation() const {}
+
+  /// Print information about the particular ordering of objects on each block
+  virtual void printBlockExecutionInformation(){}
```
```suggestion
  virtual void printBlockExecutionInformation() const {}
```

GiudGiud commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -166,6 +166,18 @@ class ThreadedElementLoopBase
+  virtual void printBlockExecutionInformation(){}
+
+  /// Keep track of which blocks were visited
+  std::set<SubdomainID> _blocks_exec_printed;
```
```suggestion
  mutable std::set<SubdomainID> _blocks_exec_printed;
```

GiudGiud committed:
Apply suggestions from code review

make block & boundary execution prints const
refs #21971
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud committed:
Add execution ordering printing to Dirac kernels
refs #21971

GiudGiud committed:
Reformatting code
Add relevant block & boundary restrictions
refs #21971

GiudGiud committed:
Simplify printing of vector of shared pointers
refs #21971

GiudGiud commented:
Might be worth waiting a bit at this point. I think there s two big tasks left

GiudGiud committed:
Refactoring of code
- more constness
refs #21971

GiudGiud committed:
Add more tests for coverage
- boundary restricted auxkernels
- element-based indicators
- nodal auxiliary bcs & kernels
- nodal UOs
refs #21971

GiudGiud committed:
Clean up finite volume flux kernel prints
- properly print header, containers are not initialized at that point so we cant rely on them
- dont give invalid subdomain id if the neighbor domain is just outside the mesh
- dont report an empty subdomain name for the dummy meshes that dont name their blocks
refs #21971

GiudGiud committed:
Only print boundary execution of auxkernels once per run
refs #21971

GiudGiud committed:
Add more tests for coverage
- nodal & element damping
- threaded and non-threaded general user objects
- bcs separate from kernels
- residual simultaneous with jacobian
refs #21971

GiudGiud committed:
Add more tests for coverage:
- interface kernesl
- dirac kernels
- fv interface kernels
- DGKernels
- domain user objects
refs #21971

lindsayad commented:
works for me. Just mention me to re-subscribe me

GiudGiud committed:
Test fixups so they pass and dont validate another one because of similar names in regex
refs #21971

GiudGiud committed:
Consolidate residual and Jacobian execution printing
- properly separate block and boundary execution prints
use same print location for BCs and interfaces, as they are tied to a boundary ID, and the logic for triggering their execution is similar
This could be confusing if:
- interface kernels are defined on a boundary that is one-sided, print would happen but execution would not. An error should have happed ealier though
refs #21971

GiudGiud committed:
Reorganize Jacobian loops to inherit from NonlinearThread
- allows a single execution print routine (original motivation)
- remove duplication of initialization code
- still allows nesting of the 4 jacobian calc types under Jacobian
refs #21971

GiudGiud commented:
I messed up a few things in the process (32 tests), but this is closer to where we want to be I think
Nesting under nonlinearthread allows me to consolidate the print routines a lot

@lindsayad

GiudGiud committed:
Fixup loop consolidation
- be more wary of which warehouses know what block we're on
- fix copy constructor for cached number
- requery the FV elemental kernels when setting mat dependencies
refs #21971

GiudGiud committed:
Add print for general user object execution directly in joinAndFinalize
I expermented with adding a separate loop for general UOs. This does not work as they depend massively on each
other. Some must be finalized, some must have their result moved into PP values & reporters values
It s just not tractable to move them cleanly out of joinAndFinalize unless..
we simply put everything in the loop and dont call joinAndFinalize on general UOs. Not great
refs #21971

GiudGiud commented:
Should be ready now. 
The 32 errors were me trying to move the general UOs into their own separate loop instead of joinAndFinalize. I should have known better.

lindsayad commented on framework/src/loops/ComputeIndicatorThread.C:
```diff
@@ -189,3 +190,43 @@ void
+  const auto & console = _fe_problem.console();
+  if (_indicator_whs.hasActiveBlockObjects(_subdomain, _tid))
+  {
+    const std::vector<std::shared_ptr<Indicator>> & indicators =
```
```suggestion
    const auto & indicators =
```

lindsayad commented on framework/src/loops/ComputeIndicatorThread.C:
```diff
@@ -189,3 +190,43 @@ void
+  }
+  if (_internal_side_indicators.hasActiveBlockObjects(_subdomain, _tid))
+  {
+    const std::vector<std::shared_ptr<InternalSideIndicator>> & indicators =
```
```suggestion
    const auto & indicators =
```

lindsayad commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -106,3 +107,19 @@ void
+void
+ComputeInitialConditionThread::printGeneralExecutionInformation() const
+{
+  const InitialConditionWarehouse & ic_wh = _fe_problem.getInitialConditionWarehouse();
```
```suggestion
  const auto & ic_wh = _fe_problem.getInitialConditionWarehouse();
```

lindsayad commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -106,3 +107,19 @@ void
+  if (_fe_problem.shouldPrintExecution(_tid) && ic_wh.hasActiveObjects())
+  {
+    const auto & console = _fe_problem.console();
+    const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
    const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeJacobianThread.C:
```diff
@@ -19,173 +19,89 @@
-    _dg_kernels(_nl.getDGKernelWarehouse()),
-    _interface_kernels(_nl.getInterfaceKernelWarehouse()),
-    _kernels(_nl.getKernelWarehouse()),
-    _tags(tags)
+  : NonlinearThread(fe_problem), _tags(tags)
```
Love it

lindsayad commented on framework/src/loops/ComputeMarkerThread.C:
```diff
@@ -108,3 +108,28 @@ void
+  if (!_fe_problem.shouldPrintExecution(_tid))
+    return;
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeMarkerThread.C:
```diff
@@ -108,3 +108,28 @@ void
+    return;
+
+  const auto & console = _fe_problem.console();
+  const std::vector<std::shared_ptr<Marker>> & markers =
```
```suggestion
  const auto & markers =
```

lindsayad commented on framework/src/loops/ComputeNodalAuxBcsThread.C:
```diff
@@ -83,6 +83,22 @@ ComputeNodalAuxBcsThread<AuxKernelType>::join(const ComputeNodalAuxBcsThread & /
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalAuxVarsThread.C:
```diff
@@ -146,6 +146,21 @@ ComputeNodalAuxVarsThread<AuxKernelType>::join(const ComputeNodalAuxVarsThread &
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalDampingThread.C:
```diff
@@ -71,3 +71,18 @@ ComputeNodalDampingThread::join(const ComputeNodalDampingThread & y)
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalKernelBCJacobiansThread.C:
```diff
@@ -141,3 +141,18 @@ void
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalKernelBcsThread.C:
```diff
@@ -97,3 +97,17 @@ void
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalKernelJacobiansThread.C:
```diff
@@ -138,3 +138,16 @@ void
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalKernelsThread.C:
```diff
@@ -98,3 +98,19 @@ void
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeNodalUserObjectsThread.C:
```diff
@@ -110,3 +109,34 @@ void
+  if (nodal_uos.size())
+  {
+    const auto & console = _fe_problem.console();
+    const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
    const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeThreadedGeneralUserObjectsThread.C:
```diff
@@ -48,3 +49,17 @@ ComputeThreadedGeneralUserObjectsThread::operator()(const GeneralUserObjectRange
+ComputeThreadedGeneralUserObjectsThread::printGeneralExecutionInformation(
+    const GeneralUserObjectRange & range) const
+{
+  // TODO: Threaded UOs dont know their thread number so this will print too often
```
This is pretty dumb that don't have a `_tid` member in this class.

GiudGiud replied:
yeah right

lindsayad commented on framework/src/loops/ComputeThreadedGeneralUserObjectsThread.C:
```diff
@@ -48,3 +49,17 @@ ComputeThreadedGeneralUserObjectsThread::operator()(const GeneralUserObjectRange
+    return;
+
+  const auto & console = _fe_problem.console();
+  const auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -22,6 +22,8 @@
 #include "DomainUserObject.h"
 #include "AuxiliarySystem.h"
+#include "MooseTypes.h"
+#include "ThreadedElementLoop.h"
```
```suggestion
```

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -321,3 +323,95 @@ void
+{
+  if (_fe_problem.shouldPrintExecution(_tid))
+  {
+    auto console = _fe_problem.console();
+    auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
    const auto & console = _fe_problem.console();
    const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -321,3 +323,95 @@ void
+      domain_interface_uos.push_back(domain_uo);
+
+  // Approximation of the number of user objects currently executing
+  int num_objects = _element_objs.size() + _domain_objs.size() + _shape_element_objs.size() +
```
```suggestion
  const auto num_objects = _element_objs.size() + _domain_objs.size() + _shape_element_objs.size() +
```

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -321,3 +323,95 @@ void
+                    side_uos.size() + shapers.size() + _internal_side_objs.size() +
+                    interface_objs.size() + domain_interface_uos.size();
+
+  auto console = _fe_problem.console();
+  auto execute_on = _fe_problem.getCurrentExecuteOnFlag();
```
```suggestion
  const auto & console = _fe_problem.console();
  const auto & execute_on = _fe_problem.getCurrentExecuteOnFlag();
```

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -321,3 +323,95 @@ void
+  if (num_objects > 0)
+  {
+    if (_blocks_exec_printed.count(_subdomain))
+      return;
```
```suggestion
      return;

```

lindsayad reviewed (changes requested):
25 files remaining

lindsayad reviewed (changes requested):
25 files remaining

lindsayad commented on framework/include/utils/MooseUtils.h:
```diff
@@ -678,18 +680,19 @@ void MaterialPropertyStorageDump(
                    const char * color = COLOR_CYAN,
-                   bool dont_indent_first_line = true);
+                   bool dont_indent_first_line = true,
+                   const std::string & post_prefix = ": ");
```
I don't love this parameter addition. Are we ever going to need to add a `post_post_prefix`?

GiudGiud replied:
no with this we control all the characters in the string. 

It's hardcoded to be a ":" because of multiapps, but [DBG]: doesnt look that good

lindsayad commented on test/tests/outputs/debug/show_execution_kernels_bcs.i:
```diff
@@ -0,0 +1,242 @@
+    grad_y = 2*y
+  []
+
+  #NeumannBC functions
```
```suggestion
  # NeumannBC functions
```

lindsayad commented on test/tests/outputs/debug/show_execution_kernels_bcs.i:
```diff
@@ -0,0 +1,242 @@
+    function = slnv
+    boundary = 'left right top bottom'
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/outputs/debug/show_execution_kernels_bcs.i:
```diff
@@ -0,0 +1,242 @@
+    vel = '0.1 0.1'
+    boundary = 'top bottom left right'
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/outputs/debug/show_execution_userobjects.i:
```diff
@@ -0,0 +1,363 @@
+    function = slnv
+    boundary = 'left right top bottom'
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/outputs/debug/show_execution_userobjects.i:
```diff
@@ -0,0 +1,363 @@
+    vel = '0.1 0.1'
+    boundary = 'top bottom left right'
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/outputs/debug/tests:
```diff
@@ -28,64 +28,64 @@
 
-     requirement = "The system shall be include the ability to display residual information for variables."
-     design = "TopResidualDebugOutput.md"
+    requirement = "The system shall be include the ability to display residual information for variables."
```
```suggestion
    requirement = "The system shall include the ability to display residual information for variables."
```

lindsayad commented on test/tests/outputs/debug/tests:
```diff
@@ -28,64 +28,64 @@
+    expect_out = "0 Linear[^*\n]*\n[^*\n]*1 Linear"
+    max_parallel = 1
+
+    requirement = "The system shall be include the ability to display residual information for variables that is limited to non-linear iterations."
```
```suggestion
    requirement = "The system shall include the ability to display residual information for variables that is limited to non-linear iterations."
```

lindsayad commented on test/tests/outputs/debug/tests:
```diff
@@ -28,64 +28,64 @@
+    expect_out = "\[DBG\]\[\d+\].*'lambda' \(SCALAR\)"
+    max_parallel = 1
+
+    requirement = "The system shall be include the ability to display residual information for scalar variables."
```
```suggestion
    requirement = "The system shall include the ability to display residual information for scalar variables."
```

lindsayad commented on test/tests/outputs/debug/tests:
```diff
@@ -97,4 +97,197 @@
+      type = RunApp
+      input = 'show_execution_adaptivity.i'
+      expect_out = '\[DBG\] indicator'
+      detail = 'each step of the adaptive refinement process, especially side-based indicators,'
```
```suggestion
      detail = 'each step of the adaptive refinement process, including side-based indicators,'
```

lindsayad commented on test/tests/outputs/debug/tests:
```diff
@@ -97,4 +97,197 @@
+      type = RunApp
+      input = 'show_execution_adaptivity.i'
+      expect_out = '\[DBG\] indicator_elem'
+      detail = 'each step of the adaptive refinement process, especially element-based indicators,'
```
```suggestion
      detail = 'each step of the adaptive refinement process, including element-based indicators,'
```

lindsayad reviewed (changes requested):
Do you expect this to affect non-debug performance at all?

GiudGiud commented:
> Do you expect this to affect non-debug performance at all?

Hopefully not.
But I did change up the Jacobian loops so maybe?

GiudGiud commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -4,16 +4,88 @@
+MOOSE parses the input file and executes numerous [Actions](actions/Action.md) which progressively
+load/build the mesh, create the variables, kernels, boundary conditions, output objects etc.
+The ordering of this process may be shown using the [!param](/Debug/show_actions) parameter.
+
+Most `Actions` are executed during initialization, but this option is also able to report on the execution of
+`Actions` during the solve.
```
```suggestion
```

GiudGiud commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -48,12 +48,25 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
    */
   virtual void compute(ResidualObject & ro) = 0;
 
+  /// Defaults to forwarding to the residual object class
```
```suggestion
  ///@{
  /// Defaults to forwarding to the residual object class
```

GiudGiud commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -48,12 +48,25 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
+  virtual void compute(IntegratedBCBase & bc);
+  virtual void compute(DGKernelBase & dg, const Elem * neighbor);
+  virtual void compute(InterfaceKernelBase & ik);
+
```
```suggestion
  ///@}

```

GiudGiud commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -77,7 +90,19 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
   /**
    * Determine the residual objects we will actually compute based on vector/matrix tag information
```
```suggestion
   * Determine the objects we will actually compute based on vector/matrix tag information
```

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2418,6 +2429,12 @@ class FEProblemBase : public SubProblem, public Restartable
 
   /// Flag used to indicate whether we are doing the uo/aux state check in execute
   bool _checking_uo_aux_state = false;
+
+  /// True if order of executions will be output
+  bool _print_execution;
```
```suggestion
```

GiudGiud commented on framework/src/loops/NonlinearThread.C:
```diff
@@ -106,15 +117,26 @@ NonlinearThread::onElement(const Elem * elem)
 
   _fe_problem.reinitMaterials(_subdomain, _tid);
 
+  if (dynamic_cast<ComputeJacobianThread *>(this))
+    if (_nl.getScalarVariables(_tid).size() > 0)
+      _fe_problem.reinitOffDiagScalars(_tid);
+
+  computeOnElement();
```
semantically it s a reinit like the materials

we would also have to duplicate in the derived classes

lindsayad replied:
ok less duplication is better

GiudGiud commented on framework/src/loops/NonlinearThread.C:
```diff
@@ -237,3 +306,99 @@ NonlinearThread::post()
+  {
+    if (_blocks_exec_printed.count(_subdomain))
+      return;
+    console << "[DBG] Ordering of " + objectType() + " Objects on block " << b_name << " ("
```
```suggestion
    console << "[DBG] Ordering of " + objectType() + " Objects on block " << block_name << " ("
```

GiudGiud commented on framework/src/loops/NonlinearThread.C:
```diff
@@ -237,3 +306,99 @@ NonlinearThread::post()
+  }
+  else if (_fe_problem.shouldPrintExecution(_tid) && num_objects == 0 &&
+           !_blocks_exec_printed.count(_subdomain))
+    console << "[DBG] No Active " + objectType() + " Objects on block " << b_name << " ("
```
```suggestion
    console << "[DBG] No Active " + objectType() + " Objects on block " << block_name << " ("
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
I ll ditch the commit then. it must be empty

GiudGiud committed:
Apply suggestions from code review

- more references when retrieving objects
- fix grammar tests requirement
- better casts of objects to residual elem
- fix grouped docstrings
refs #21971
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1430237/) on cab55e2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21586/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9eedce8060b5450a00d114861901084b554e58d2`

GiudGiud committed:
Address CIVET test failures:
- dont hide the compute routines
- missing page for debugging in downstream apps
- bad doco links
- inconsistent print if constructs are not on domain 0 (should we print anyway?)
refs #21971

GiudGiud commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -48,12 +48,25 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
    */
   virtual void compute(ResidualObject & ro) = 0;
 
+  /// Defaults to forwarding to the residual object class
```
done

GiudGiud commented:
@lindsayad going to resolve the suggestions manually because it didnt auto-resolve on merging them. I checked that i didnt miss any though

lindsayad commented:
I went through and resolved all but 2 it looks like

GiudGiud commented:
well that explains the clicking not working
thanks!

moosebuild commented:
Job [Test timings](https://civet.inl.gov/job/1430398/) on 9e0381e wanted to post the following:

View timings [here](https://mooseframework.inl.gov/docs/PRs/21586/timing)

This comment will be updated on new commits.

GiudGiud commented:
I have tracked down the app failure to this commit 23c14883d8b6cc27e323fb6ae3c539a40f6276e8
it doesnt make much sense. I ll see if my new debugging tool can help me figure it ^^

tophmatthews commented:
I'm not sure what's going on with the bison failure. The test is for simple 3d thermal expansion verification test. The strange thing is that the AD version works just fine, but the nonad version hits a `DIVERGED_NANORINF`. It uses `decomposition_method = EigenSolution` and small strain, and nothing else too crazy. It can be simplified by turning the temperature into an aux variable, but I fear the fact nonad isn't working and AD is indicates something deeper...

GiudGiud commented:
Found the problem. I'm ashamed how long it took

GiudGiud committed:
Double down on removing the NONE execute_on flag during residual/jacobian evaluations
refs #21971

GiudGiud commented:
We should be unstuck now. This was a good thing overall, i think we're in a better place with that last commit. 

Down for another round of review!

lindsayad reviewed (commented):
This is ready as far as I'm concerned. I will approve this once I can help get next unstuck and we can get everything here green

moosebuild commented:
Job [Python 3.8](https://civet.inl.gov/job/1441521/) on 89024e5 : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1441522/) on 89024e5 : invalidated by @lindsayad
